runtime/doc/options.txt

```diff
@@ -4390,6 +4390,27 @@ A jump table for the options with a short description can be found at |Q_op|.
 	    |nobody         |  3 nobody     |  0 nobody     |3   nobody
 	    |there          |  4 there      |  1 there      |  1 there
 
+				*'numberformat'* *'nuf'*
+'numberformat' 'nuf'	string	(default: empty)
+			local to window
+	When this option is not empty, it determines the content of the number
+	column. The format of this option is like that of 'statusline'. The
+	variables *g:nuflnum* and *g:nufrnum* will be set to the to be printed line
```

Comment:
can/should they be `v:` vars? Similar to `v:event`, maybe a common container like `v:redraw` is a good pattern. So these would be `v:redraw.lnum` and `v:redraw.relnum`.

Related to performance, accessing vimscript dicts (`g:`, `v:`) is slow, but we might be able to optimize `v:` vars because they are readonly.

Reply:
Sounds reasonable, also left another comment regarding performance that would require adding new 'statusline' expression items.

Reply:
With the below benchmark by running `time nvim --clean -S test.vim`(commenting out the relevant line):
```vim
set number nowrap
set numbercolumn=%l
"set numbercolumn=%{v:redraw_lnum}
"set numberformat=%{nuflnum}

norm ia
norm yy9999P
let i = 0
while i < 9999
	let i += 1
	norm j
  redraw!
endwhile
quit!
```
There is no way of telling(with this rudimentary test) whether using a read-only `v:` scope variable is actually faster than a `g:` scope variable (currently, because you say it might still be able to be optimized). The lowest runtime alternates between the g:/v: builds in my testing.

The `set numbercolumn=%l` variant, without an `%{}` eval block, **is** measurably faster. Presumably the overhead in setting up the eval environment(not sure that's proper jargon) is the determining factor here, making differences(if any) in variable scope insignificant.

So there might be merit in adding new `'statusline`' expression items specifically for `'numbercolumn'`. It feels kind of iffy since they wouldn't be useful for the other statusline-format options but we might even go so far as to provide things like hex/thousands separator (relative)line number items as well. That way most users wouldn't need an eval block in their `'numbercolumn'` I think, but we can still set variables for those that do. Different colors for for odd/even line numbers is one mentioned use case that comes to mind where evaluating the line number would be necessary.

> > We could also add new format expression items containing the to be printed (relative)number.
 
> Rather not. Let's optimize `v:` access instead (if it's not already optimized).

Considering the above, would you still rather not? Care to elaborate?

Reply:
Let's stick with `v:` for now. Could also overload `%l` as you discuss in https://github.com/neovim/neovim/pull/20621#discussion_r994894476